subsys/emds/emds.c

```diff
@@ -0,0 +1,327 @@
+			ssize_t len = emds_flash_write(&emds_flash,
+						       ch->entry.id, ch->entry.data, ch->entry.len);
+			if (len < 0) {
+				LOG_ERR("Write dynamic entry: (%d) error (%d).",
+					ch->entry.id, len);
+			}
+			if (len != ch->entry.len) {
+				LOG_ERR("Write dynamic entry: (%d) failed (%d:%d).",
+					ch->entry.id, ch->entry.len, len);
+			}
+		}
+
+		emds_ready = false;
+
+		k_sched_unlock();
+
+		/* Unlock all interrupts */
+		irq_unlock(store_key);
+
+		if (app_store_cb) {
+			app_store_cb();
+		}
+	}
+}
+
+K_THREAD_DEFINE(emds_tid, CONFIG_EMDS_THREAD_STACK_SIZE, emds_handler, NULL,
+		NULL, NULL, K_PRIO_COOP(CONFIG_EMDS_THREAD_PRIORITY), 0, 0);
+
+static int emds_fs_init(void)
+{
+	int rc;
+	const struct flash_area *fa;
+	uint32_t sector_cnt = 1;
+	struct flash_sector hw_fs;
+	size_t emds_sector_size;
+	size_t emds_storage_size = 0;
+	uint16_t cnt = 0;
+
+	rc = flash_area_open(FLASH_AREA_ID(emds_storage), &fa);
+	if (rc) {
+		return rc;
+	}
+
+	rc = flash_area_get_sectors(FLASH_AREA_ID(emds_storage), &sector_cnt,
+						  &hw_fs);
+	if (rc == -ENODEV) {
+		return rc;
+	} else if (rc != 0 && rc != -ENOMEM) {
+		k_panic();
+	}
+
+	emds_sector_size = hw_fs.fs_size;
+
+	if (emds_sector_size > UINT16_MAX) {
+		return -EDOM;
+	}
+
+	while (cnt < CONFIG_EMDS_SECTOR_COUNT) {
+		emds_storage_size += emds_sector_size;
+		if (emds_storage_size > fa->fa_size) {
+			break;
+		}
+		cnt++;
+	}
+
+	emds_flash.sector_size = emds_sector_size;
+	emds_flash.sector_cnt = cnt;
+	emds_flash.offset = fa->fa_off;
+
+	return emds_flash_init(&emds_flash, fa->fa_dev_name);
+}
+
+
+static int emds_entries_size(uint32_t *size)
+{
+	size_t block_size = emds_flash.flash_params->write_block_size;
+	int entries = 0;
+
+	*size = 0;
+
+	STRUCT_SECTION_FOREACH(emds_entry, ch) {
+		*size += NRFX_CEIL_DIV(ch->len, block_size) * block_size;
+		*size += NRFX_CEIL_DIV(emds_flash.ate_size, block_size) * block_size;
+		entries++;
+	}
+
+	struct emds_dynamic_entry *ch;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		*size += NRFX_CEIL_DIV(ch->entry.len, block_size) * block_size;
+		*size += NRFX_CEIL_DIV(emds_flash.ate_size, block_size) * block_size;
+		entries++;
+	}
+
+	return entries;
+}
+
+int emds_init(emds_store_cb_t cb)
+{
+	int rc;
+
+	if (emds_initialized) {
+		LOG_DBG("Already initialized");
+		return -EALREADY;
+	}
+
+	rc = emds_fs_init();
+	if (rc) {
+		return rc;
+	}
+
+	sys_slist_init(&emds_dynamic_entries);
+
+	app_store_cb = cb;
+
+	emds_initialized = true;
+
+	LOG_DBG("Emergency Data Storage initialized.");
+
+	return 0;
+}
+
+int emds_entry_add(const struct emds_entry *entry)
+{
+	struct emds_dynamic_entry *ch;
+	struct emds_dynamic_entry *item;
+
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+		if (ch->entry.id == entry->id) {
+			return -EINVAL;
+		}
+	}
+
+	item = k_malloc(sizeof(struct emds_dynamic_entry));
```

Comment:
Does it have the sense to use heap if there is no allocated memory freeing? Seems there is no necessity in heap. It mostly requires an `item` field within the dynamic entry structure. IMHO

Reply:
If emds does not create this then, then the caller has to create this entry and keep it forever. This is done because the caller can call this function and then forget about this. We have not created a remove function so a memory free function would have been in that. The hole thing is that this should be kept until it is rebooted, because of the prepare function. This way the caller of add cannot do anything wrong by clearing the entry, and making the list having empty entries.

Reply:
It is possible to encapsulate `item` within `struct emds_dynamic_entry`. Doing that it will allow getting rid of all code related to heap usage (potential less size of feature since allocator\deallocator of the heap is not linked). I guess this task for future component improvement is not a blocking reason at the current moment.

Reply:
Created JIRA ticket for this NCSDK-15171